src/physics/utilities/boundary_condition.cpp

```diff
@@ -48,18 +48,19 @@ void BoundaryCondition::project(FiniteElementState& state) const
     std::transform(tdofs.begin(), tdofs.end(), dof_list.begin(),
                    [&space = std::as_const(state.space())](int tdof) { return space.VDofToDof(tdof); });
 
-    if (component_ == -1) {
-      // If it contains all components, project the vector
-      auto vec_coef = std::get_if<std::shared_ptr<mfem::VectorCoefficient>>(&coef_);
-      SLIC_ASSERT_MSG(vec_coef,
-                      "Essential boundary condition contained all components but had a non-vector coefficient.");
-      state.gridFunc().ProjectCoefficient(**vec_coef, dof_list);  // Deref ptr-to-result, then shared_ptr
+    // the only reason to store a VectorCoefficient is to act on all components
+    if (std::holds_alternative<std::shared_ptr<mfem::VectorCoefficient>>(coef_)) {
+      auto vec_coef = std::get<std::shared_ptr<mfem::VectorCoefficient>>(coef_);
```

Comment:
```cpp
if (auto vec_coef = std::get_if<std::shared_ptr<mfem::VectorCoefficient>>(&coef))
```

Reply:
how about this? I like the symmetry of the if/else that comes with using `std::get<>`
```cpp
    if (is_vector_valued(coef_)) {
      auto vec_coef = std::get<std::shared_ptr<mfem::VectorCoefficient>>(coef_);
      state.gridFunc().ProjectCoefficient(*vec_coef, dof_list);
    } else {
      auto scalar_coef = std::get<std::shared_ptr<mfem::Coefficient>>(coef_);
      state.gridFunc().ProjectCoefficient(*scalar_coef, dof_list, std::max(0, component_));
    }
```
these two functions get added under the definition of `GeneralCoefficient`:
```cpp
bool is_scalar_valued(GeneralCoefficient coef) {
  return std::holds_alternative<std::shared_ptr<mfem::Coefficient>>(coef);
}

bool is_vector_valued(GeneralCoefficient coef) {
  return std::holds_alternative<std::shared_ptr<mfem::VectorCoefficient>>(coef);
}
```